src/pages/audio-rewards-page/store/sagas.ts

```diff
@@ -260,38 +295,61 @@ function* watchUpdateHCaptchaScore() {
   })
 }
 
-function* pollForChallenges(): any {
-  const pollingFreq = remoteConfigInstance.getRemoteVar(
-    IntKeys.CHALLENGE_REFRESH_INTERVAL_MS
-  )
-  if (pollingFreq) {
-    const chan = yield call(doEvery, pollingFreq, function* () {
-      yield put(fetchUserChallenges())
-    })
-    yield take(reset.type)
-    chan.close()
-  }
-}
+function* visibilityPollingDaemon(action: Action, delayTimeMs: number) {
+  // Set up daemon that will watch for browser into focus and refetch notifications
+  // as soon as it goes into focus
+  const visibilityChannel = eventChannel(emitter => {
+    if (NATIVE_MOBILE) {
+      // The focus and visibitychange events are wonky on native mobile webviews,
+      // so poll for visiblity change instead
+      let lastHidden = true
+      setInterval(() => {
+        if (!document.hidden && lastHidden) {
+          emitter(true)
+        }
+        lastHidden = document.hidden
+      }, 500)
+    } else {
+      document.addEventListener('visibilitychange ', () => {
+        if (!document.hidden) {
+          emitter(true)
+        }
+      })
+    }
+    return () => {}
+  })
+  yield fork(function* () {
+    while (true) {
+      yield take(visibilityChannel)
+      yield put(action)
+    }
+  })
 
-function* pollForBalance(): any {
-  const pollingFreq = remoteConfigInstance.getRemoteVar(
-    IntKeys.REWARDS_WALLET_BALANCE_POLLING_FREQ_MS
+  // Set up daemon that will poll for notifications every 10s if the browser is
+  // in the foreground
+  let isBrowserInBackground = false
+  document.addEventListener(
+    'visibilitychange',
+    () => {
+      if (document.hidden) {
+        isBrowserInBackground = true
+      } else {
+        isBrowserInBackground = false
+      }
+    },
+    false
   )
-  if (pollingFreq) {
-    const chan = yield call(doEvery, pollingFreq, function* () {
-      yield put(getBalance())
-    })
-    yield take(reset.type)
-    chan.close()
-  }
-}
 
-function* watchRefreshUserChallenges() {
-  yield takeEvery(refreshUserChallenges.type, pollForChallenges)
+  while (true) {
+    if (!isBrowserInBackground || isElectron()) {
+      yield put(action)
+    }
+    yield delay(delayTimeMs)
+  }
 }
 
-function* watchRefreshUserBalance() {
-  yield takeEvery(refreshUserBalance.type, pollForBalance)
+function* userChallengePollingDaemon() {
+  yield* visibilityPollingDaemon(fetchUserChallenges(), 5000)
```

Comment:
todo: Make poll time configurable

Reply:
optimizely would be great

Reply:
+1 for optimizely. was discussing this with saliou a while back - for the sake of load, it might make sense to have two polling intervals, one for when we're on the $AUDIO page and care more about more frequent updates, and one when we're sitting elsewhere and can tolerate more latency.